### PR TITLE
Add in tutorial and recipes for sparse transfer learning with Ultralytics YOLOv3

### PR DESCRIPTION
**Pull request by markurtz**

### CODE REVIEW

**Review by dhuangnm — COMMENTED** [review]
Mark, should we also include the new tutorial under the Tutorials section in this README page?

https://github.com/neuralmagic/sparseml/tree/main/integrations/ultralytics-yolov3/README.md

**Comment by markurtz** [comment]
> Mark, should we also include the new tutorial under the Tutorials section in this README page?
> 
> https://github.com/neuralmagic/sparseml/tree/main/integrations/ultralytics-yolov3/README.md

Yes, definitely, good catch Dan

File: integrations/ultralytics-yolov3/recipes/yolov3-spp.transfer_learn_pruned.md [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+Adjust the script command for your GPU device setup. 
+Ultralytics supports both DataParallel and DDP.
+
+*script command:*
+
+```bash
+python train.py \
+    --data voc.yaml \
+    --cfg ../models/yolov3-spp.lrelu.yaml \
+    --weights DOWNLOADED_PATH \
```
Comment (bfineran) [inline integrations/ultralytics-yolov3/recipes/yolov3-spp.transfer_learn_pruned.md:155]: can we have a note or comment on where to get the weights from (and the SparseZoo stub for the weights to reproduce)

File: integrations/ultralytics-yolov3/recipes/yolov3-spp.transfer_learn_pruned_quantized.md [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+Adjust the script command for your GPU device setup. 
+Ultralytics supports both DataParallel and DDP.
+
+*script command:*
+
+```bash
+python train.py \
+    --data voc.yaml \
+    --cfg ../models/yolov3-spp.lrelu.yaml \
+    --weights DOWNLOADED_PATH \
```
Comment (bfineran) [inline integrations/ultralytics-yolov3/recipes/yolov3-spp.transfer_learn_pruned_quantized.md:224]: can we have a note or comment on where to get the weights from (and the SparseZoo stub for the weights to reproduce)

**Review by bfineran — COMMENTED** [review]
LGTM overall, I think having numbered sections for the sections where readers are expected to do something would make the doc easier to parse and give a sense of overall progress

File: integrations/ultralytics-yolov3/recipes/yolov3-spp.transfer_learn_pruned.md [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+    start_epoch: *pruning_start_epoch
+    end_epoch: *pruning_end_epoch
+    update_frequency: *pruning_update_frequency
+---
+
+# YOLOv3-SPP Pruned Transfer Learning
+
+This recipe transfer learns from a sparse, [YOLOv3-SPP](https://arxiv.org/abs/1804.02767) model.
+It was originally tested on the VOC dataset and achieved 0.84 mAP@0.5. 
+More information on the run can be found in the Weights and Biases project.
```
Comment (jeanniefinks) [inline integrations/ultralytics-yolov3/recipes/yolov3-spp.transfer_learn_pruned.md:132]: add the link to the Weights and Biases project?

File: integrations/ultralytics-yolov3/recipes/yolov3-spp.transfer_learn_pruned_quantized.md [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+      - model.25
+      - model.26
+      - model.27
+---
+
+# YOLOv3-SPP Pruned Quantized Transfer Learning
+
+This recipe transfer learns from a sparse-quantized, [YOLOv3-SPP](https://arxiv.org/abs/1804.02767) model.
+It was originally tested on the VOC dataset and achieved 0.838 mAP@0.5. 
+More information on the run can be found in the Weights and Biases project.
```
Comment (jeanniefinks) [inline integrations/ultralytics-yolov3/recipes/yolov3-spp.transfer_learn_pruned_quantized.md:201]: same comment as initial identical one:
add the link to the Weights and Biases project? (even though you mention it later, it would seem natural to link to the first reference here esp if you are telling the user that more info can found there...)

File: integrations/ultralytics-yolov3/tutorials/yolov3_sparse_transfer_learning.md [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+limitations under the License.
+-->
+
+# Sparse Transfer Learning With YOLOv3
+
+This tutorial shows how Neural Magic sparse models simplify the sparsification process by offering pre-sparsified models for transfer learning onto other datasets.
+
+## Overview
+
+Neural Magic’s ML team created sparsified models that allow anyone to plug in their data and leverage pre-sparsified models from the SparseZoo on top of Ultralytics’ robust training pipelines.
```
Comment (jeanniefinks) [inline integrations/ultralytics-yolov3/tutorials/yolov3_sparse_transfer_learning.md:23]:
```suggestion
Neural Magic’s ML team creates sparsified models that allow anyone to plug in their data and leverage pre-sparsified models from the SparseZoo on top of Ultralytics’ robust training pipelines. 
```

File: integrations/ultralytics-yolov3/tutorials/yolov3_sparse_transfer_learning.md [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+    ```
+   
+You are now ready to setup the data for training.
+
+## Setting Up The Data
+
+Note: If using your custom data, the Ultralytics repo contains a walk-through for [training custom data](https://github.com/ultralytics/yolov3/wiki/Train-Custom-Data). 
+Otherwise, setup scripts for both [VOC](https://cs.stanford.edu/~roozbeh/pascal-context/) and [COCO](https://cocodataset.org/#home) can be found under the [yolov3/data/scripts path](https://github.com/ultralytics/yolov3/tree/master/data/scripts).
+
+1) For this tutorial, run the VOC setup script with the following command from the root of the yolov3 repository:
```
Comment (jeanniefinks) [inline integrations/ultralytics-yolov3/tutorials/yolov3_sparse_transfer_learning.md:89]:
```suggestion
1) For this tutorial, run the VOC setup script with the following command from the root of the `yolov3` repository:
```

File: integrations/ultralytics-yolov3/tutorials/yolov3_sparse_transfer_learning.md [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+   - Pruned Quantized transfer learning, achieves 0.838 mAP@0.5 on the VOC dataset:
+        ```bash
+        python train.py --data voc.yaml --cfg ../models/yolov3-spp.lrelu.yaml --weights DOWNLOADED_PATH --hyp data/hyp.finetune.yaml --recipe ../recipes/yolov3-spp.transfer_learn_pruned_quantized.md
+        ```
+   - Baseline transfer learning, achieves 0.86 mAP@0.5 on the VOC dataset:
+        ```bash
+        python train.py --data voc.yaml --cfg ../models/yolov3-spp.lrelu.yaml --weights DOWNLOADED_PATH --hyp data/hyp.finetune.yaml
+        ```
+
+    Weights and Biases is very useful for comparing across the different runs, the epochs vs mAP@0.5 graph is supplied below:
```
Comment (jeanniefinks) [inline integrations/ultralytics-yolov3/tutorials/yolov3_sparse_transfer_learning.md:145]:
```suggestion
    Weights and Biases is very useful for comparing across the different runs; the epochs vs mAP@0.5 graph is supplied below:
```

File: integrations/ultralytics-yolov3/tutorials/yolov3_sparse_transfer_learning.md [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+    
+    <img src="https://raw.githubusercontent.com/neuralmagic/sparseml/main/integrations/ultralytics-yolov3/tutorials/images/transfer-learning-wandb-chart.png" width="960px" style="border: 2px solid #000000;" />
+
+Afterward, you will have a sparse model trained on the VOC dataset almost ready for inference.
+The training command creates a runs directory under the yolov3 repository directory.
```
Comment (jeanniefinks) [inline integrations/ultralytics-yolov3/tutorials/yolov3_sparse_transfer_learning.md:150]:
```suggestion
The training command creates a runs directory under the `yolov3` repository directory. 
```

File: integrations/ultralytics-yolov3/tutorials/yolov3_sparse_transfer_learning.md [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+|-- README.md
+`-- setup_integration.sh
+```
+
+## Exporting for Inference
+
+This step loads a checkpoint file of the best weights measured on the validation set, and converts it into the more common inference formats. 
+Then, you can run the file through a compression algorithm to reduce its deployment size and run it in an inference engine such as DeepSparse.
+
+The best.pt file contains a checkpoint of the best weights measured on the validation set.
```
Comment (jeanniefinks) [inline integrations/ultralytics-yolov3/tutorials/yolov3_sparse_transfer_learning.md:184]:
```suggestion
The `best.pt` file contains a checkpoint of the best weights measured on the validation set. 
```

File: integrations/ultralytics-yolov3/tutorials/yolov3_sparse_transfer_learning.md [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+`-- setup_integration.sh
+```
+
+## Exporting for Inference
+
+This step loads a checkpoint file of the best weights measured on the validation set, and converts it into the more common inference formats. 
+Then, you can run the file through a compression algorithm to reduce its deployment size and run it in an inference engine such as DeepSparse.
+
+The best.pt file contains a checkpoint of the best weights measured on the validation set. 
+These weights can be loaded into the train.py and test.py scripts now.
```
Comment (jeanniefinks) [inline integrations/ultralytics-yolov3/tutorials/yolov3_sparse_transfer_learning.md:185]:
```suggestion
These weights can be loaded into the `train.py` and `test.py` scripts now. 
```

File: integrations/ultralytics-yolov3/tutorials/yolov3_sparse_transfer_learning.md [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+    ```
+    |-- exp
+    |-- exp1
+    |-- weights
+    |   |-- best.pt
+    |   |-- best.onnx
+    |   `-- last.pt
+    `-- ...
+    ```
+2) Now you can run the .onnx file through a compression algorithm to reduce its deployment size and run it in ONNX-compatible inference engines such as [DeepSparse](https://github.com/neuralmagic/deepsparse).
```
Comment (jeanniefinks) [inline integrations/ultralytics-yolov3/tutorials/yolov3_sparse_transfer_learning.md:204]:
```suggestion
2) Now you can run the `.onnx` file through a compression algorithm to reduce its deployment size and run it in ONNX-compatible inference engines such as [DeepSparse](https://github.com/neuralmagic/deepsparse).
```

File: integrations/ultralytics-yolov3/tutorials/yolov3_sparse_transfer_learning.md [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+    </a>
+</div>
+
+## Need Help?
+
+For Neural Magic Support, sign up or log in to get help with your questions in our Tutorials channel: [Discourse Forum](https://discuss.neuralmagic.com/) and/or [Slack](https://join.slack.com/t/discuss-neuralmagic/shared_invite/zt-q1a1cnvo-YBoICSIw3L1dmQpjBeDurQ). 
+
+## Downloading and Preparing a Pre-pruned Model
+
+First you need to download the sparsified models from the [SparseZoo](https://github.com/neuralmagic/sparsezoo).
```
Comment (jeanniefinks) [inline integrations/ultralytics-yolov3/tutorials/yolov3_sparse_transfer_learning.md:48]:
```suggestion
First, you need to download the sparsified models from the [SparseZoo](https://github.com/neuralmagic/sparsezoo).
```

File: integrations/ultralytics-yolov3/tutorials/yolov3_sparse_transfer_learning.md [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+
+2) Once the desired checkpoint has downloaded, it must be reset for training again.
+    The `utility.py` script within YOLOv3 repository is used for this.
+    First change into the `yolov3` directory using `cd yolov3`.
+    Next run the following command to prepare the checkpoint for training:
+    ```bash
+    python utility.py strip /PATH/TO/DOWNLOADED/WEIGHTS.pt
+    ```
+   
+You are now ready to setup the data for training.
```
Comment (jeanniefinks) [inline integrations/ultralytics-yolov3/tutorials/yolov3_sparse_transfer_learning.md:82]:
```suggestion
You are now ready to set up the data for training.
```

**Review by jeanniefinks — COMMENTED** [review]
Done w/ edit suggestions

File: integrations/ultralytics-yolov3/recipes/yolov3-spp.transfer_learn_pruned.md (hunk 1 above)
Comment (markurtz) [reply integrations/ultralytics-yolov3/recipes/yolov3-spp.transfer_learn_pruned.md:155]: yes, good point, adding

File: integrations/ultralytics-yolov3/recipes/yolov3-spp.transfer_learn_pruned_quantized.md (hunk 2 above)
Comment (markurtz) [reply integrations/ultralytics-yolov3/recipes/yolov3-spp.transfer_learn_pruned_quantized.md:224]: yes, good point, adding

File: integrations/ultralytics-yolov3/recipes/yolov3-spp.transfer_learn_pruned.md (hunk 3 above)
Comment (markurtz) [reply integrations/ultralytics-yolov3/recipes/yolov3-spp.transfer_learn_pruned.md]: good point, moved the statement to under the weights and biases section

File: integrations/ultralytics-yolov3/recipes/yolov3-spp.transfer_learn_pruned_quantized.md [hunk 14]
```diff
@@ -0,0 +1,241 @@
...
+      - model.21
+      - model.22
+      - model.23
+      - model.24
+      - model.25
+      - model.26
+      - model.27
+---
+
+# YOLOv3-SPP Pruned Quantized Transfer Learning
```
Comment (jeanniefinks) [inline integrations/ultralytics-yolov3/recipes/yolov3-spp.transfer_learn_pruned_quantized.md:197]:
```suggestion
# YOLOv3-SPP Pruned-Quantized Transfer Learning
```

File: integrations/ultralytics-yolov3/tutorials/yolov3_sparse_transfer_learning.md [hunk 15]
```diff
@@ -0,0 +1,215 @@
...
+    The `utility.py` script within YOLOv3 repository is used for this.
+    First change into the `yolov3` directory using `cd yolov3`.
+    Next run the following command to prepare the checkpoint for training:
+    ```bash
+    python utility.py strip /PATH/TO/DOWNLOADED/WEIGHTS.pt
+    ```
+   
+You are now ready to set up the data for training.
+
+## Setting Up The Data
```
Comment (jeanniefinks) [inline integrations/ultralytics-yolov3/tutorials/yolov3_sparse_transfer_learning.md:84]:
```suggestion
## Setting Up the Data
```

File: integrations/ultralytics-yolov3/tutorials/yolov3_sparse_transfer_learning.md [hunk 16]
```diff
@@ -0,0 +1,215 @@
...
+Without the proper recipes, the zeroed weights will slowly become dense as the model is trained further.
+The recipes are specific to the sparsification type, so the training command will differ based on if you are transfer learning the pruned quantized, pruned, or baseline model.
+
+1) Select the proper command to run based on the sparsification type of the model you downloaded earlier.
+   Change directory into the `yolov3` folder to use the `train.py` script.
+   - Pruned transfer learning, achieves 0.84 mAP@0.5 on the VOC dataset:
+        ```bash
+        python train.py --data voc.yaml --cfg ../models/yolov3-spp.lrelu.yaml --weights DOWNLOADED_PATH --hyp data/hyp.finetune.yaml --recipe ../recipes/yolov3-spp.transfer_learn_pruned.md
+        ```
+   - Pruned Quantized transfer learning, achieves 0.838 mAP@0.5 on the VOC dataset:
```
Comment (jeanniefinks) [inline integrations/ultralytics-yolov3/tutorials/yolov3_sparse_transfer_learning.md:136]:
```suggestion
   - Pruned quantized transfer learning, achieves 0.838 mAP@0.5 on the VOC dataset:
```

File: integrations/ultralytics-yolov3/tutorials/yolov3_sparse_transfer_learning.md [hunk 17]
```diff
@@ -0,0 +1,215 @@
...
+        ```bash
+        python train.py --data voc.yaml --cfg ../models/yolov3-spp.lrelu.yaml --weights DOWNLOADED_PATH --hyp data/hyp.finetune.yaml
+        ```
+
+    Weights and Biases is very useful for comparing across the different runs; the epochs vs mAP@0.5 graph is supplied below:
+    
+    <img src="https://raw.githubusercontent.com/neuralmagic/sparseml/main/integrations/ultralytics-yolov3/tutorials/images/transfer-learning-wandb-chart.png" width="960px" style="border: 2px solid #000000;" />
+
+Afterward, you will have a sparse model trained on the VOC dataset almost ready for inference.
+The training command creates a runs directory under the `yolov3` repository directory.
```
Comment (jeanniefinks) [inline integrations/ultralytics-yolov3/tutorials/yolov3_sparse_transfer_learning.md:150]:
```suggestion
The training command creates a `runs` directory under the `yolov3` repository directory. 
```

**Review by bfineran — DISMISSED** [review]

**Review by jeanniefinks — CHANGES_REQUESTED** [review]
Done reviewing. Minor changes.

**Review by jeanniefinks — APPROVED** [review]
